imap/http_proxy.c

```diff
@@ -1243,3 +1261,212 @@ EXPORTED int http_proxy_copy(struct backend *src_be, struct backend *dest_be,
+
+    /*
+     * Send client request to backend:
+     *
+     * - Change CONNECT to HTTP/1.1 Upgrade
+     * - Add/append-to Via: header
+     * - Use all cached end-to-end headers from client
+     */
+    prot_printf(be->out, "GET %s %s\r\n"
+                         "Host: %s\r\n"
+                         "User-Agent: Cyrus/%s\r\n",
+                txn->req_uri->path, HTTP_VERSION,
+                be->hostname, CYRUS_VERSION);
+    write_forwarding_hdrs(&be_txn, txn->req_hdrs, txn->req_line.ver,
+                          https ? "https" : "http");
+    prot_puts(be->out, "Connection: Upgrade\r\n");
+    hdr = spool_getheader(txn->req_hdrs, ":protocol");
+    if (hdr && *hdr) {
+        prot_printf(be->out, "Upgrade: %s\r\n", *hdr);
+    }
+    prot_printf(be->out, "Sec-WebSocket-Key: %s\r\n",
+                "Q3lydXMgSFRUUCBQcm94eQ==");  // "Cyrus HTTP Proxy" b64-encoded
+    spool_enum_hdrcache(txn->req_hdrs, &write_cachehdr, &be_txn);
+    prot_puts(be->out, "\r\n");
+    prot_flush(be->out);
+    buf_free(&be_txn.buf);
+
+    /* Read response from backend */
+    resp_body.flags = 0;
+
+    r = http_read_response(be, METH_GET, &code,
+                           &resp_hdrs, &resp_body, &txn->error.desc);
+    if (r || (resp_body.flags & BODY_CLOSE)) {
+        proxy_downserver(be);
+        goto done;
+    }
+
+    /* Send response to client */
+    if (code == 101) {
+        code = 200;
+        txn->flags.te = TE_CHUNKED;
+    }
+    http_err = http_status_to_code(code);
+    send_response(txn, http_err, resp_hdrs, &resp_body.payload);
+
+    log_proxy_request(http_err, txn, resp_hdrs, &resp_body);
+
+ done:
+    buf_free(&resp_body.payload);
+    if (resp_hdrs) spool_free_hdrcache(resp_hdrs);
+
+    return r;
+}
+
+/*
+ * Check an array of streams for input.
+ *
+ * Input from serverin is sent to clientout.
+ * If serverout is non-NULL:
+ *   - input from clientin is sent to serverout.
+ *   - returns -1 if clientin or serverin closed, otherwise returns 0.
+ * If serverout is NULL:
+ *   - returns 1 if input from clientin is pending, otherwise returns 0.
+ */
+EXPORTED int http_proxy_check_input(struct http_connection *conn,
+                                    ptrarray_t *pipes,
+                                    unsigned long timeout_sec)
+{
+    struct protgroup *protin = conn->pgin;
+    struct protgroup *protout = NULL;
+    struct timeval timeout = { timeout_sec, 0 };
+    struct protstream *clientin = conn->pin;
+    struct protstream *clientout = conn->pout;
+    struct protstream *serverout = NULL;
+    struct transaction_t *txn = NULL;
+    int n, ret = 0;
+
+    protgroup_reset(protin);
+    protgroup_insert(protin, clientin);
+
+    for (n = 0; n < ptrarray_size(pipes); n++) {
+        txn = ptrarray_nth(pipes, n);
+        protgroup_insert(protin, txn->be->in);
+
+        if (txn->flags.ver < VER_2) {
+            /* stream client input directly to server */
+            serverout = txn->be->out;
```

Comment:
This feels like, if there's more than one txn in `pipes`, it could end up replacing `serverout` a bunch of times.  Does that need to be guarded against?

Reply:
Good question, but for HTTP/1.1 there can only be one txn at a time due to the nature of the protocol.  HTTP/2+ is multiplexed, but we can't stream directly to the server here.  We need to return to the caller so that we can parse the framing and determine what stream we are dealing with.  I'll add a comment to the code to explain this.